File: c/src/ntt_mont_asm.c
#include <bigint.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>

#define NTT_MOD (998244353ULL)
#define NTT_ROOT (3ULL)
/* MONTGOMERY_R_INV = (-NTT_MOD)^-1 mod 2^64 */
#define MONTGOMERY_R_INV (17450252288407896063ULL)

/*
 * montgomery_mul(a, b, mod) computes (a * b * R^-1) mod `mod`
 * where R = 2^64.
 *
 * It requires `a` and `b` to be in Montgomery form: a = A*R mod mod.
 * The constant MONTGOMERY_R_INV must be (-mod)^-1 mod 2^64.
 */
static inline uint64_t montgomery_mul(uint64_t a, uint64_t b, uint64_t mod) {
    uint64_t result;
    /*
     * 1. Compute T = a * b (128-bit product in rdx:rax)
     * 2. Compute m = (T_lo * MONTGOMERY_R_INV) mod 2^64
     * 3. Compute T + m * mod. The low 64 bits are guaranteed 0.
     * 4. The result is the high 64 bits, minus mod if it overflowed >= mod.
     */
    __asm__ (
        "mulq %2\n\t"               // rdx:rax = a * b
        "movq %%rax, %%r8\n\t"        // r8 = T_lo
        "movq %%rdx, %%r9\n\t"        // r9 = T_hi
        
        "imulq %3, %%rax\n\t"         // rax = m = T_lo * mod_inv mod 2^64
        "mulq %4\n\t"                 // rdx:rax = m * mod
        
        "addq %%r8, %%rax\n\t"        // rax = T_lo + (m * mod)_lo  (always 0, sets carry)
        "adcq %%r9, %%rdx\n\t"        // rdx = T_hi + (m * mod)_hi + carry  (this is (T + m*mod)/2^64)
        
        "movq %%rdx, %%rax\n\t"       // rax = result
        "subq %4, %%rdx\n\t"          // rdx = result - mod
        "cmovnc %%rdx, %%rax\n\t"     // if (result >= mod) result = result - mod
        : "=&a"(result)               // %0
        : "0"(a),                     // %1 (same as %0, rax)
          "r"(b),                     // %2
          "r"(MONTGOMERY_R_INV),      // %3
          "r"(mod)                    // %4
        : "rdx", "r8", "r9", "cc"
    );
    return result;
}

static uint64_t mod_pow_asm(uint64_t base, uint64_t exp, uint64_t mod) {
    // 1 in Montgomery form is R mod MOD. R = 2^64. 2^64 % 998244353 = 932051910
    uint64_t result = 932051910ULL;
    uint64_t b = base % mod;
    
    while (exp > 0) {
        if (exp & 1) {
            result = montgomery_mul(result, b, mod);
        }
        b = montgomery_mul(b, b, mod);
        exp >>= 1;
    }
    return result;
}

static void ntt_inplace_asm(uint64_t *a, size_t n, uint64_t mod, uint64_t root, bool inverse) {
    for (size_t i = 1, j = 0; i < n; i++) {
        size_t bit = n >> 1;
        for (; j & bit; bit >>= 1) j ^= bit;
        j ^= bit;
        if (i < j) {
            uint64_t t = a[i]; a[i] = a[j]; a[j] = t;
        }
    }
    
    size_t len = 2;
    while (len <= n) {
        uint64_t wlen = mod_pow_asm(root, (mod - 1) / len, mod);
        if (inverse) wlen = mod_pow_asm(wlen, mod - 2, mod);
        
        uint64_t w = 932051910ULL; // 1 in Montgomery form
        for (size_t j = 0; j < len / 2; j++) {
            for (size_t i = 0; i < n; i += len) {
                uint64_t u = a[i + j];
                uint64_t v = montgomery_mul(a[i + j + len / 2], w, mod);
                a[i + j] = u + v;
                if (a[i + j] >= mod) a[i + j] -= mod;
                a[i + j + len / 2] = u + mod - v;
                if (a[i + j + len / 2] >= mod) a[i + j + len / 2] -= mod;
            }
            w = montgomery_mul(w, wlen, mod);
        }
        len <<= 1;
    }
    
    if (inverse) {
        uint64_t mont_n = montgomery_mul(n % mod, 299560064ULL, mod);
        uint64_t inv_n = mod_pow_asm(mont_n, mod - 2, mod);
        for (size_t i = 0; i < n; i++) {
            a[i] = montgomery_mul(a[i], inv_n, mod);
        }
    }
}

static void *xrealloc(void *ptr, size_t size) {
    void *p = realloc(ptr, size);
    if (!p && size > 0) abort();
    return p;
}

BigUInt *biguint_mul_ntt_mont_asm(const BigUInt *a, const BigUInt *b) {
    if (biguint_is_zero(a) || biguint_is_zero(b)) {
        return biguint_new();
    }
    
    size_t n = 1;
    while (n < a->len + b->len - 1) n <<= 1;
    size_t result_len = a->len + b->len - 1;
    
    uint64_t *fa = calloc(n, sizeof(uint64_t));
    uint64_t *fb = calloc(n, sizeof(uint64_t));
    
    if (!fa || !fb) {
        free(fa); free(fb);
        return NULL;
    }
    
    /*
     * R = 2^64 mod NTT_MOD. We need this to encode inputs into Montgomery form.
     * 2^64 = 18446744073709551616.
     * 18446744073709551616 % 998244353 = 136407672
     */
    /* R^2 mod NTT_MOD. Used to convert normal -> Montgomery */
    const uint64_t R2_MOD = 299560064ULL;
    
    // Encode inputs to Montgomery form: a_mont = a * R mod NTT_MOD
    // We achieve this by montgomery_mul(a, R2_MOD) = a * R^2 * R^-1 = a * R
    for (size_t i = 0; i < a->len; i++) fa[i] = montgomery_mul(a->words[i] % NTT_MOD, R2_MOD, NTT_MOD);
    for (size_t i = 0; i < b->len; i++) fb[i] = montgomery_mul(b->words[i] % NTT_MOD, R2_MOD, NTT_MOD);
    
    // The NTT_ROOT also needs to be in Montgomery form
    uint64_t mont_root = montgomery_mul(NTT_ROOT, R2_MOD, NTT_MOD);
    
    ntt_inplace_asm(fa, n, NTT_MOD, mont_root, false);
    ntt_inplace_asm(fb, n, NTT_MOD, mont_root, false);
    
    for (size_t i = 0; i < n; i++) {
        fa[i] = montgomery_mul(fa[i], fb[i], NTT_MOD);
    }
    
    ntt_inplace_asm(fa, n, NTT_MOD, mont_root, true);
    
    // Decode from Montgomery form: montgomery_mul(x, 1) = x * 1 * R^-1 = x * R^-1
    // which cancels the R factor.
    for (size_t i = 0; i < n; i++) {
        fa[i] = montgomery_mul(fa[i], 1, NTT_MOD);
    }
    
    BigUInt *res = biguint_new();
    if (!res) {
        free(fa); free(fb);
        return NULL;
    }
    
    res->words = xrealloc(NULL, result_len * sizeof(uint64_t));
    if (!res->words) {
        free(fa); free(fb);
        free(res);
        return NULL;
    }
    
    for (size_t i = 0; i < result_len; i++) {
        res->words[i] = fa[i];
    }
    res->len = result_len;
    res->cap = result_len;
    
    free(fa); free(fb);
    biguint_normalize(res);
    
    return res;
}

File: rust/src/ntt_mont_asm.rs
use crate::bigint::{next_power_of_two, BigUInt};

const NTT_MOD: u64 = 998244353;
const NTT_ROOT: u64 = 3;
const MONTGOMERY_R_INV: u64 = 17450252288407896063; // (-998244353)^-1 mod 2^64
const R2_MOD: u64 = 299560064;

#[inline]
fn montgomery_mul(a: u64, b: u64) -> u64 {
    let t = (a as u128) * (b as u128);
    let m = (t as u64).wrapping_mul(MONTGOMERY_R_INV);
    let m_mod = (m as u128) * (NTT_MOD as u128);
    
    // (t + m_mod) is mathematically guaranteed to be divisible by 2^64
    let mut result = ((t + m_mod) >> 64) as u64;
    
    if result >= NTT_MOD {
        result -= NTT_MOD;
    }
    result
}

// mod_mul removed, use montgomery_mul.

fn mod_pow_asm(mut base: u64, mut exp: u64) -> u64 {
    let mut result = montgomery_mul(1, R2_MOD); // 1 in montgomery form
    // Note: base is assumed to be in Montgomery form natively.
    
    while exp > 0 {
        if exp & 1 == 1 {
            result = montgomery_mul(result, base);
        }
        base = montgomery_mul(base, base);
        exp >>= 1;
    }
    result
}

fn ntt_inplace(a: &mut [u64], mod_val: u64, root: u64, inverse: bool) {
    let n = a.len();

    let mut j = 0;
    for i in 1..n {
        let mut bit = n >> 1;
        while j & bit != 0 {
            j ^= bit;
            bit >>= 1;
        }
        j ^= bit;
        if i < j {
            a.swap(i, j);
        }
    }

    let mut len = 2;
    while len <= n {
        let wlen = mod_pow_asm(root, (mod_val - 1) / len as u64);
        let wlen = if inverse {
            mod_pow_asm(wlen, mod_val - 2)
        } else {
            wlen
        };

        for i in (0..n).step_by(len) {
            let mut w = montgomery_mul(1, R2_MOD);
            for j in 0..len / 2 {
                let u = a[i + j];
                let v = montgomery_mul(a[i + j + len / 2], w);
                a[i + j] = u + v;
                if a[i + j] >= mod_val {
                    a[i + j] -= mod_val;
                }
                a[i + j + len / 2] = u + mod_val - v;
                if a[i + j + len / 2] >= mod_val {
                    a[i + j + len / 2] -= mod_val;
                }
                w = montgomery_mul(w, wlen);
            }
        }
        len <<= 1;
    }

    if inverse {
        let inv_n = mod_pow_asm(montgomery_mul(n as u64, R2_MOD), mod_val - 2);
        for x in a.iter_mut() {
            *x = montgomery_mul(*x, inv_n);
        }
    }
}

pub fn biguint_mul_ntt_mont_asm(a: &BigUInt, b: &BigUInt) -> BigUInt {
    if a.is_zero() || b.is_zero() {
        return BigUInt::new();
    }

    let n = next_power_of_two(a.words().len() + b.words().len() - 1);
    let result_len = a.words().len() + b.words().len() - 1;

    let mut fa: Vec<u64> = vec![0; n];
    let mut fb: Vec<u64> = vec![0; n];

    for i in 0..a.words().len() {
        fa[i] = montgomery_mul(a.words()[i] % NTT_MOD, R2_MOD);
    }
    for i in 0..b.words().len() {
        fb[i] = montgomery_mul(b.words()[i] % NTT_MOD, R2_MOD);
    }

    let mont_root = montgomery_mul(NTT_ROOT, R2_MOD);

    ntt_inplace(&mut fa, NTT_MOD, mont_root, false);
    ntt_inplace(&mut fb, NTT_MOD, mont_root, false);

    for i in 0..n {
        fa[i] = montgomery_mul(fa[i], fb[i]);
    }

    ntt_inplace(&mut fa, NTT_MOD, mont_root, true);

    for i in 0..n {
        fa[i] = montgomery_mul(fa[i], 1);
    }

    let mut result = Vec::with_capacity(result_len);
    for i in 0..result_len {
        result.push(fa[i]);
    }

    BigUInt::from_slice(&result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mul_small() {
        let a = BigUInt::from_u64(12345);
        let b = BigUInt::from_u64(67890);
        let c = biguint_mul_ntt_mont_asm(&a, &b);
        let expected = BigUInt::from_u64(12345 * 67890);
        assert_eq!(c, expected);
    }
}
